### DISCUSSION TITLE
[KeepHiddenConstructorAttribute] A Attribute for default hiding and disallowing constructor for factory method

### DISCUSSION THREAD

**Discussion by Shadowblitz16** [dotnet/csharplang#2726] [Language Ideas]
Lets say I have classes like so..
```
public class A
{
    protected A()
    {
    }
    public static T MakeA<T>() where T : A, new()
    {
        return new T()
    }
} 
public class B : A
{
}
```
and I wanted users to only use the factory pattern like so..
```
B b = A.MakeA<B>()
```
this works but as of now users can simply call
```
B b = new B() { //somthing here };
```
which defeats the purpose of incapsulation for users that don't want to have to write 
```
protected B()
{
}
```
all the time.

I propose that a attribute be made to hide constructors from intellisense and disallow them being called from places other then T and the class they are in

**Comment by yaakov-h** [comment]
What's to stop me from then writing:

```csharp
class C
{
    public static B MakeB() => Make<B>();
    static T Make<T> where T : new() => new T();
}
```

?

Also, this sounds like a great use-case for a custom Roslyn Analyzer.

**Comment by khm1600** [comment]
@yaakov-h
> ~~CS0310 'B' must be a non-abstract type with a public parameterless constructor in order to use it as parameter 'T' in the generic type or method 'C.Make<T>()'~~

**Comment by yaakov-h** [comment]
@khm1600 [hmmm?](https://sharplab.io/#v2:C4LgbgpgTglgZgTwDQBMQGoA+B6bACAYQHsUI84oiBbPAeQAUBYAKBYAEBmPNgJjwEEWAbxZ4xeAA6VgEAMYyUAgBQBKUeJHNxeAL7qxnbgEYAbHgAqeALIBDANYR+AHnMA+VXgDuAC2hlLIAJIeAB2EJ6q+nia2tpsAOyh4RaqANxResw6eOxcvHgAQniBgsyamSy41gh4siQQLOx8BMJRhmymhdb2EAUeALyu3Q5OBe4q6VriHWaWtiNuHj5+FsVJESp4g+spEyw6QA===)

**Comment by khm1600** [comment]
@yaakov-h 
I misunderstood the OP, I guess.

**Comment by Shadowblitz16** [comment]
@yaakov-h it wouldn't. 
you can still make your own constructor methods you just can't call the constructor directly

**Comment by JakenVeina** [comment]
```cs
public static T MakeA<T>() where T : A, new()
{
    return new T()
}
```
```cs
where T : A, new()
```
You are literally imposing a requirement on all subclasses that they MUST implement a `public` parameterless constructor. Kinda renders this use-case moot.

Sounds more like what you want is to just hide the constructor from intelllisense, which still strikes me as an over-zealous goal.

**Comment by Shadowblitz16** [comment]
don't classes by default have a public parameterless constructor?
this would not only hide the constructor from intellisense but throw a error if it is called outside of a generic parameter.

**Comment by theunrepentantgeek** [comment]
@Shadowblitz16 asked
> don't classes by default have a public parameterless constructor?

Only if you don't declare any constructors yourself.

@JakenVeina suggested
> Sounds more like what you want is to just hide the constructor from intelllisense, which still strikes me as an over-zealous goal.

If this is the goal, use the attribute that already exists. Just decorate the constructor with:
```
[EditorBrowsable(EditorBrowsableState.Never)]
```
Note - doesn't apply within the same project, only when referencing the assembly from elsewhere.

**Comment by Thaina** [comment]
@Shadowblitz16 

> don't classes by default have a public parameterless constructor?

No that is only for `struct`

**Comment by Shadowblitz16** [comment]
@theunrepentantgeek I also didn't want the user to be allowed to call it.

**Comment by Thaina** [comment]
I think you are hitting the limit of factory pattern And I can't think of anything that cannot be a loophole in C#

I think the actual point is, we need factory to orchestrate some pipeline that class always needed to do. But it couldn't be call from constructor. If that was the case I was propose #2722

Or if you have more requirement than that and need to use factory creation at all cost. I think maybe you could made it by stripping all of protected member from `A`. Make `A` into empty sealed class. And move all those into interface for struct, then the `MakeA<T>` would require a struct that implement your interface. This way you couldn't inherit A and cannot create it anywhere else

**Comment by yaakov-h** [comment]
I'm just going to re-raise this point:

> this sounds like a great use-case for a custom Roslyn Analyzer.

If you want to forbid something that the language permits, this is exactly the purpose of custom Analyzers.

**Comment by Shadowblitz16** [comment]
The new() constraint supporting visible protected constructors would be a better option.
